napari/layers/_data_protocols.py

```diff
@@ -0,0 +1,129 @@
+"""This module holds Protocols that layer.data objects are expected to provide.
+"""
+from __future__ import annotations
+
+from types import GeneratorType
+from typing import (
+    TYPE_CHECKING,
+    Any,
+    List,
+    Optional,
+    Sequence,
+    Tuple,
+    TypeVar,
+    Union,
+)
+
+import numpy as np
+from typing_extensions import Protocol, runtime_checkable
+
+from .utils.layer_utils import compute_multiscale_level_and_corners
+
+_T = TypeVar('_T')
+Shape = Tuple[int, ...]
+ListOrTuple = Union[List[_T], Tuple[_T, ...], np.ndarray]
+_OBJ_NAMES = set(dir(Protocol))
+_OBJ_NAMES.update({'__annotations__', '__dict__', '__weakref__'})
+
+if TYPE_CHECKING:
+    from ..types import DTypeLike
+
+
+def _raise_protocol_error(obj: Any, protocol: type):
+    """Raise a more helpful error when required protocol members are missing."""
+    needed = set(dir(protocol)).union(protocol.__annotations__) - _OBJ_NAMES
+    missing = needed - set(dir(obj))
+    message = (
+        f"Object of type {type(obj).__name__!r} does not implement "
+        f"{protocol.__name__!r} Protocol.\n"
+        f"Missing methods: {missing!r}"
+    )
+    raise TypeError(message)
+
+
+@runtime_checkable
+class LayerDataProtocol(Protocol):
+    """A Protocol that all layer.data needs to support.
+
+    WIP: Shapes.data may be an execption.
```

Comment:
Are you imagining that shapes data would be a `Sequence[LayerDataProtocol]`?

Reply:
Yeah, shapes is `Sequence[LayerDataProtocol]` ... there it's a little less problematic, since it's _always_ a list container of arrays, whereas with Image, it could be an array or a sequence of arrays, and you don't know until you check.  We could consider creating a custom container that obeys the `LayerDataProtocol` as well, but that's what I was punting on :)

Reply:
makes a lot of sense. I definitely agree that `Sequence[LayerDataProtocol]` for shapes makes more sense than a custom container :)